src/login/logind-core.c

```diff
@@ -602,3 +602,39 @@ bool manager_is_docked_or_external_displays(Manager *m) {
 
         return false;
 }
+
+bool manager_is_on_external_power(Manager *m) {
+        int r;
+
+        /* For backwards compatibility we require a valid action to be set in
+         * order for us to take this property into account */
+        if (m->handle_lid_switch_ep == _HANDLE_ACTION_INVALID)
```

Comment:
I can't say I like that: someone will see a function `manager_is_on_external_power` without reading the sources and they will think it does what the name says. This check must be moved out of the function or the name must be changed.